api/v1beta1/azuremachine_validation.go

```diff
@@ -378,3 +392,37 @@ func ValidateDiagnostics(diagnostics *Diagnostics, fieldPath *field.Path) field.
 
 	return allErrs
 }
+
+// ValidateConfidentialCompute validates the configuration options when the machine is a Confidential VM.
+func ValidateConfidentialCompute(managedDisk *ManagedDiskParameters, profile *SecurityProfile, fieldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	var securityEncryptionType SecurityEncryptionType
+
+	if managedDisk != nil && managedDisk.SecurityProfile != nil {
+		securityEncryptionType = managedDisk.SecurityProfile.SecurityEncryptionType
+	}
+
+	if profile != nil && securityEncryptionType != "" {
+		if profile.EncryptionAtHost != nil && *profile.EncryptionAtHost && securityEncryptionType == SecurityEncryptionTypeDiskWithVMGuestState {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("EncryptionAtHost"), profile.EncryptionAtHost,
+				"EncryptionAtHost cannot be set to 'true' when securityEncryptionType is set to DiskWithVMGuestState"))
+		}
+
+		if profile.VirtualizedTrustedPlatformModule == "" || profile.VirtualizedTrustedPlatformModule == VirtualizedTrustedPlatformModulePolicyDisabled {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("VirtualizedTrustedPlatformModule"), profile.VirtualizedTrustedPlatformModule,
+				fmt.Sprintf("VirtualizedTrustedPlatformModule should be set to '%s' when securityEncryptionType is defined",
+					VirtualizedTrustedPlatformModulePolicyEnabled)))
+		}
+
+		if securityEncryptionType == SecurityEncryptionTypeDiskWithVMGuestState {
+			if profile.SecureBoot == "" || profile.SecureBoot != SecureBootPolicyEnabled {
+				allErrs = append(allErrs, field.Invalid(fieldPath.Child("SecureBoot"), profile.SecureBoot,
+					fmt.Sprintf("SecureBoot should be set to '%s' when securityEncryptionType is set to '%s'",
```

Comment:
same comment as above, I'm sure about this one :)

you could argue CAPZ should enforce strong defaults but if you force secure boot and a user has custom unsigned modules, you can break workloads for no reason.

Reply:
Great feedback, in the same incorrect sense, I blindly followed the Azure Terraform Provider [implementation](https://github.com/hashicorp/terraform-provider-azurerm/blob/7a3afb9d59a4cca6b1c410a3cd1a2d9db1d92d85/internal/services/compute/linux_virtual_machine_resource.go#L540-L546)... 

I will test it out (just for completeness sake) and revert this change as well.

Reply:
if we do confirm terraform is wrong, maybe we can open some issues :) if it was correct before, thanks for confirming and helping me learn something ;)

Reply:
Trying to create a `ConfidentialVM` with `SecurityEncryptionType: DiskWithVMGuestState`, `vTpmEnabled: true` and `secureBootEnabled: false` fails with the following error:

```shell
> az vm create \
  --resource-group aResourceGroup \
  --name myConfidentialVM \
  --generate-ssh-keys \
  --size Standard_DC4as_v5 \
  --admin-username aUsername \
  --admin-password aPassword \
  --enable-vtpm true \
  --image "Canonical:0001-com-ubuntu-confidential-vm-focal:20_04-lts-cvm:latest" \
  --public-ip-sku Standard \
  --security-type ConfidentialVM \
  --os-disk-security-encryption-type DiskWithVMGuestState \
  --enable-secure-boot false \

ERROR: {"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.","details":[{"code":"BadRequest","message":"{\r\n  \"error\": {\r\n    \"code\": \"BadRequest\",\r\n    \"message\": \"Confidential VM or Virtual Machine Scale Set VM with 'managedDisk.securityProfile.securityEncryptionType' set as DiskWithVMGuestState should have both values of securityProfile.uefiSettings.secureBootEnabled and securityProfile.uefiSettings.vtpmEnabled set to true.\"\r\n  }\r\n}"}]}}
```